lib/puppet/provider/exec.rb

```diff
@@ -4,71 +4,71 @@
-              end
-              environment[env_name] = value
-            else
-              warning _("Cannot understand environment setting %{setting}") % { setting: setting.inspect }
-            end
-          end
-        end
-
-        # Ruby 2.1 and later interrupt execution in a way that bypasses error
-        # handling by default. Passing Timeout::Error causes an exception to be
-        # raised that can be rescued inside of the block by cleanup routines.
-        #
-        # This is backwards compatible all the way to Ruby 1.8.7.
-        Timeout::timeout(resource[:timeout], Timeout::Error) do
-          # note that we are passing "false" for the "override_locale" parameter, which ensures that the user's
-          # default/system locale will be respected.  Callers may override this behavior by setting locale-related
-          # environment variables (LANG, LC_ALL, etc.) in their 'environment' configuration.
-          output = Puppet::Util::Execution.execute(command, :failonfail => false, :combine => true,
-                                  :uid => resource[:user], :gid => resource[:group],
-                                  :override_locale => false,
-                                  :custom_environment => environment,
-                                  :sensitive => sensitive)
-        end
-        # The shell returns 127 if the command is missing.
-        if output.exitstatus == 127
-          raise ArgumentError, output
-        end
 
-      end
-    rescue Errno::ENOENT => detail
-      self.fail Puppet::Error, detail.to_s, detail
+    # Ruby 2.1 and later interrupt execution in a way that bypasses error
+    # handling by default. Passing Timeout::Error causes an exception to be
+    # raised that can be rescued inside of the block by cleanup routines.
+    #
+    # This is backwards compatible all the way to Ruby 1.8.7.
+    Timeout::timeout(resource[:timeout], Timeout::Error) do
+      # If we're running a command that's meant to be a check on whether we should run
+      # our actual command (e.g. like a command passed to the :onlyif or :unless properties),
+      # then we should not set the cwd when executing the check's corresponding command.
+      cwd = check ? nil : resource[:cwd]
+      cwd ||= Dir.pwd
+
+      # note that we are passing "false" for the "override_locale" parameter, which ensures that the user's
+      # default/system locale will be respected.  Callers may override this behavior by setting locale-related
+      # environment variables (LANG, LC_ALL, etc.) in their 'environment' configuration.
+      output = Puppet::Util::Execution.execute(
+        command,
+        :failonfail => false,
+        :combine => true,
+        :cwd => cwd,
+        :uid => resource[:user], :gid => resource[:group],
+        :override_locale => false,
+        :custom_environment => environment,
```

Comment:
I know it's considered "normal" Ruby style to leave parens off for a method call with no arguments, but I always prefer to see them.

Reply:
Yeah, I can make this call more readable.